File: vol2/ch17/LeetCode315-2.py
from sortedcontainers import SortedList
class Solution:
    def countSmaller(self, nums: List[int]) -> List[int]:
        n=len(nums)
        sl=SortedList()
        ans=[0]*n
        for i in range(n-1,-1,-1):
            ans[i]=sl.bisect_left(nums[i])
            sl.add(nums[i])
        return ans

File: vol2/ch17/LeetCode315-2.cpp
class Solution {
public:
    vector<int> countSmaller(vector<int>& nums) {
        int n=nums.size();
        vector<int> ans;
        vector<int> tmp;
        ans.push_back(0);
        tmp.push_back(nums[n-1]);
        for(int i=n-2;i>=0;i--) {
            auto it=lower_bound(tmp.begin(),tmp.end(),nums[i]);
            ans.push_back(it-tmp.begin());
            tmp.insert(it,nums[i]);
        }
        reverse(ans.begin(),ans.end());
        return ans;
    }
};

